data/tutorials/wf_03_metaprogramming.md

```diff
@@ -0,0 +1,454 @@
+to drive the preprocessing). Consider the following simple preprocessor which
+replaces the string `"World"` by the string `"Universe"`, here in the form of a
+shell script:
+
+```shell
+$ cat preprocessor.sh
+#!/bin/sh
+
+sed 's/World/Universe/g' $1
+```
+
+Compiling a classic "Hello, World!" program with this option
+would alter the compilation:
+
+```shell
+$ cat hello.ml
+print_endline "Hello, World!";;
+
+$ ocamlopt -pp preprocessor.sh hello.ml
+$ ./a.out
+Hello, Universe!
+```
+
+### Preprocessing with Dune
+
+Dune's build system has a specific stanza to apply preprocessing to files.
+The full documentation can be found
+[here](https://dune.readthedocs.io/en/stable/concepts.html#preprocessing-specification),
+and should serve as a reference. In this section, we only give a few examples.
+
+The stanza to apply preprocessing on the source file is `(preprocess (action
+(<action>)))`. The `<action>` part can be any user-defined action, such as the
+call to an external program, as specified by `(system <program>)`.
+
+Putting it all together, the following `dune` file would rewrite the corresponding
+module files using our previously written `preprocessor.sh`:
+
+```dune
+(executable
+ (name main)
+ (preprocess
+  (action
+   (system "./preprocessor.sh %{input-file}"))))
+```
+
+### The Limits of Manipulating Text Files
+
+The complexities of a programming language syntax makes it very hard to
+manipulate text in a way that is tied to the syntax. Suppose for instance that,
+similarly to the previous example, you want to rewrite all occurences of "World"
+by "Universe," but _inside the OCaml strings of the program_ only. It is quite
+involved and requires a good knowledge of the OCaml syntax to do so, as there
+are several delimiters for strings (such as `{| ...|}`) and line breaks, or
+comments could get in the way...
+
+Consider another example. Suppose you have defined a type, and you want to
+generate a serializer at compile time from this specific type to an encoding of
+it in a `json` format, such as `Yojson` (see
+[here](#why-are-ppxes-especially-useful-in-ocaml) for the reasons it has to be
+generated at compile-time). This serialization code could be written by a
+preprocessor, which would look for the type in the file and serialize it
+differently depending on the type structure; that is, whether it is a variants
+type, a record type, the structure of its subtypes...
+
+But, to understand the type's structure, the preprocessor needs to parse
+this information. Given that parsing OCaml syntax is complex, it is not
+something that we want to do every time we write a new
+preprocessor.
+
+The solution to this is to use PPXs, preprocessors that run on the result of
+the file parsing.
+
+## PPXs
+
+PPxs are a different kind of preprocessors—one that do not run on the textual
+source code, but rather on the parsing result: the Abstract Syntax Tree (AST),
+or more precisely in OCaml, the Parsetree. In order to understand PPXs well, we
```

Comment:
This suggestion clarifies it and works well.

Reply:
In my mind, both the OCaml Parstree and Typedtree were ASTs. After some researching, it seems the word "parsetree" is usually used to refer to CSTs rather than ASTs, see for instance [this stackoverflow question](https://stackoverflow.com/questions/5026517/whats-the-difference-between-parse-trees-and-abstract-syntax-trees-asts).

But, from this "terminological research", it is clear that in the case of OCaml, AST and Parsetree are synonym and I will go for your suggestion!